File: 04-program-structure/c04s01-c/storage/db/net/sql.c
#include <stdio.h>
#include "storage.h"

static struct aa_sql_connection_t singleton;
static void get_data(void);

struct aa_sql_connection_t *create_sql_connection(void)
{
    printf("create_sql_connection()\n");
    if(singleton.get_data != get_data){
      singleton.get_data = get_data;
    }
  
    return &singleton;
}

static void get_data(void){
  printf("get_data() - getting data from Sql conn\n");
}

File: src/04-program-structure/c04s03-go/storage/db/sql/sql.go
package sql

import (
	"04-program-structure/c04s03-go/storage"
	"fmt"
)

func Factory() storage.Storer {
	fmt.Println("Creating SQL storage engine")
	return &SqlStorer{}
}

type SqlStorer struct {
	// .. fields
}

func (r *SqlStorer) GetData(b []byte) (int, error) {
	fmt.Println("GetData from Sql Storage engine")
	return len(b), nil
}
func (r *SqlStorer) Write(b []byte) (int, error) {
	fmt.Println("Writing data to Sql Storage engine")
	return len(b), nil
}
func (r *SqlStorer) Read(b []byte) (int, error) {
	fmt.Println("Reading data from Sql Storage engine")
	return len(b), nil
}
